### PR TITLE
Introducing the StepAction CRD

### PR DESCRIPTION
**Pull request by chitrangpatel**
This PR introduces the StepAction CRD in `v1alpha1`. This is not the complete CRD. As we add support for `params`, `results` etc. we will introduce them in follow-up PRs.

It is the first of many implementation PRs as described in issue  https://github.com/tektoncd/pipeline/issues/7259.

# Changes

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Introduced the StepAction CRD as designed in https://github.com/tektoncd/community/blob/main/teps/0142-enable-step-reusability.md. The implementation is still ongoing; feature is not yet functional.
```
/kind feature

### CODE REVIEW

File: pkg/apis/pipeline/v1alpha1/stepaction_defaults.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+	"context"
+
+	"knative.dev/pkg/apis"
+)
+
+var _ apis.Defaultable = (*StepAction)(nil)
+
+// SetDefaults implements apis.Defaultable
+func (s *StepAction) SetDefaults(ctx context.Context) {
+	// s.Spec.SetDefaults(ctx)
```
Comment (Yongxuanzhang) [inline pkg/apis/pipeline/v1alpha1/stepaction_defaults.go:26]: Do we need this?

File: pkg/apis/pipeline/v1alpha1/stepaction_conversion.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	"context"
+	"fmt"
+
+	"knative.dev/pkg/apis"
+)
+
+var _ apis.Convertible = (*StepAction)(nil)
+
+// ConvertTo implements apis.Convertible
+func (s *StepAction) ConvertTo(ctx context.Context, to apis.Convertible) error {
```
Comment (Yongxuanzhang) [inline pkg/apis/pipeline/v1alpha1/stepaction_conversion.go:26]: Do we need the conversions? It basically converts with itself

Comment (chitrangpatel) [reply pkg/apis/pipeline/v1alpha1/stepaction_conversion.go:26]: Oh yes. good question. We probably don't need this. I may have introduced it when I was trying to also have it in v1. I will fix it.

File: pkg/apis/pipeline/v1alpha1/stepaction_defaults.go (hunk 1 above)
Comment (chitrangpatel) [reply pkg/apis/pipeline/v1alpha1/stepaction_defaults.go:26]: Probably not now. Once we add params and results, we will. I can remove it for now.

File: cmd/webhook/main.go [hunk 3]
```diff
@@ -169,6 +171,13 @@ func newConversionController(ctx context.Context, cmw configmap.Watcher) *contro
 		// conversions to and from all types.
 		// "Zygotes" are the supported versions.
 		map[schema.GroupKind]conversion.GroupKindConversion{
+			v1alpha1.Kind("StepAction"): {
```
Comment (Yongxuanzhang) [inline cmd/webhook/main.go:174]: We can add the conversion webhook when we need it?

**Review by Yongxuanzhang — APPROVED** [review]

File: cmd/webhook/main.go (hunk 3 above)
Comment (chitrangpatel) [reply cmd/webhook/main.go:174]: We need this for validation web hook. However, I removed the `Zygotes`.

File: sa.yaml
Comment (jerop) [inline sa.yaml:1]: please add this file in the examples folder, and give a clear name 🙏🏾

Comment (chitrangpatel) [reply sa.yaml:1]: I'm going to remove it for now. It makes more sense to add it when we introduce the feature flag and support for its consumption in Tasks. 

I accidentally commit the file while ensuring that the changes to the validation web hook etc. worked.

Removing for now.

Comment (chitrangpatel) [reply sa.yaml:1]: Done!

File: config/300-stepaction.yaml [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    - tekton-pipelines
+  scope: Namespaced
+  conversion:
+    strategy: Webhook
+    webhook:
+      conversionReviewVersions: ["v1beta1", "v1"]
+      clientConfig:
+        service:
+          name: tekton-pipelines-webhook
+          namespace: tekton-pipelines
```
Comment (jerop) [inline config/300-stepaction.yaml:61]: do we need this conversion?

Comment (chitrangpatel) [reply config/300-stepaction.yaml]: Good catch. I think we only need it when we promote the CRD from v1alpha1 to v1beta1 and then v1.
Fixed.

**Review by jerop — APPROVED** [review]
/lgtm

**Comment by chitrangpatel** [comment]
> Sorry for being a bad cop here 🙏 As excited as this is (introducing a significant new concept in the project) did not follow the diversity requirements:
> 
> **The Tekton community promotes company diversity as a best practice for pull request. This means that, where possible, one of the reviewers of a pull request and the author should be affiliated to different organizations.**
> 
> This policy was introduced after a discussion at the Governance WG:
> 
> [tektoncd/community#829](https://github.com/tektoncd/community/pull/829)
> 
> /cc @tektoncd/core-maintainers

Agreed! Let's not repeat the same mistake in the follow up PRs. 
For now, is there any way we can undo the change so that others can review this?

**Comment by jerop** [comment]
We were just eager to review in a timely manner. We will keep an eye on that guidance going forward. Thank you for pointing it out @pritidesai! 

@pritidesai @tektoncd/core-maintainers if you have any feedback on this change, @chitrangpatel can address in a follow up PR. 

We would also appreciate any help in reviewing future PRs as soon as you can as we are targeting v0.54 release next month 🙏🏾

**Comment by Yongxuanzhang** [comment]
> Sorry for being a bad cop here 🙏 As excited as this is (introducing a significant new concept in the project) did not follow the diversity requirements:
> 
> **The Tekton community promotes company diversity as a best practice for pull request. This means that, where possible, one of the reviewers of a pull request and the author should be affiliated to different organizations.**
> 
> This policy was introduced after a discussion at the Governance WG:
> 
> [tektoncd/community#829](https://github.com/tektoncd/community/pull/829)
> 
> /cc @tektoncd/core-maintainers

Thanks @pritidesai for the reminder, we should definitely follow the rules!

**Comment by jerop** [comment]
> 0.53 is not out yet and the focus right now is on one last pending PR 🙏 :

> https://github.com/tektoncd/pipeline/pulls?q=is%3Aopen+is%3Apr+milestone%3A%22Pipelines+v0.53+LTS%22

That's right, but that one remaining PR has feedback from several maintainers who can merge it when feedback is addressed -- happy to take a look at that PR if it helps

Started reviewing this PR targeting v0.54 because this is a huge chunk of work split over many pull requests, so would like to get some head start to make it in time instead of rushing near the deadline

**Pull request merged**